### ISSUE TITLE
Support for extending Micrometer tags

### ISSUE THREAD

**Issue by batigoal82** [resilience4j/resilience4j#509]
I want to be able to extend the tags added to a gauge in the micrometer library.

But currently all classes responsible for metrics tags have a private constructor and thus cannot be easily extended.

* `TaggedRetryMetrics`
* `TaggedCirciutBreakerMetrics`
* `TaggedBulkheadMetrics`
* `TaggedRateLimiterMetrics`

Could you make the classes extendable or add a way I could provide the bean as in `RetryMetricsAutoConfiguration` myself that is not a closed class?

 Thanks!

**Comment by RobWin** [comment]
Would you like to create a PR?

**Comment by batigoal82** [comment]
yes, I will create a PR.

**Comment by RobWin** [comment]
Do you want to add global tags, which should be added to every gauge or only a specific one?

**Comment by batigoal82** [comment]
> Do you want to add global tags, which should be added to every gauge or only a specific one?

I would like to add specific ones.

**Comment by RobWin** [comment]
Ok, is it a static value or a dynamic value which must be discovered at runtime?
I'm asking if it make sense to provide an option to configure static tags via the config file.

**Comment by batigoal82** [comment]
> Ok, is it a static value or a dynamic value which must be discovered at runtime?
> I'm asking if it make sense to provide an option to configure static tags via the config file.

It is a dynamic value which is discovered at runtime. I like your idea regarding configurable static tags although I have a doubt that they would be used.

**Comment by RobWin** [comment]
The beans `TaggedRetryMetrics` in `RetryMetricsAutoConfiguration` already use the annotation `ConditionalOnMissingBean`. Isn't it already possible to overwrite the bean (even without your PR)?

Could you use a `MeterFilter` to add a additional tag to certain metrics? See https://micrometer.io/docs/concepts#_transforming_metrics and https://docs.spring.io/spring-boot/docs/current/reference/html/production-ready-metrics.html#production-ready-metrics-per-meter-properties

**Comment by siepkes** [comment]
Were any API's for this added?

In my case I have a larger application which consists of modules and was looking to add a simple tag with something like `module=<MODULE NAME>` for metrics used in specific a module. So while the tag name is set at compile time it's not a static tag in the sense that it needs to be added globally to all metrics.

**Comment by RobWin** [comment]
Hi,
no, we did not add anything yet. A PR is welcome. 
Do  you need the `module` tag to filter for all CircuitBreakers instances which are used in a certain module?

As a workaround you could use an instance name like: `moduleName#instanceName` and use a regex-based filter in Grafana.

**Comment by siepkes** [comment]
More then willing to create a PR!

Yes, the idea of the module tag was to be able to filter on all instances in a certain module. Thanks for the workaround tip! Though for my inner perfectionist (and for the sake of keeping things simple and uniform for the team) I prefer a "proper" solution :wink:. I think the workaround will also be harder to use with Prometheus' AlertManager.

I haven't look indepth what the best API for this would be. Do you have an idea what would be acceptable for you folks? The first thing that pops in to my mind would be something like:

```
TaggedRetryMetrics
		.ofRetryRegistry(retryRegistry)
		.bindTo(registry)
		.addTags(tags);

```

**Comment by siepkes** [comment]
Upon closer inspection of the API I noticed quick enough the proposed API won't work. What would work is adding a telescoping constructor like below to a class like `TaggedBulkheadMetrics`: 

```java
public static TaggedBulkheadMetrics ofBulkheadRegistry(Tags tags, MetricNames names, BulkheadRegistry bulkheadRegistry) { }
```

Would that be acceptable?

**Comment by RobWin** [comment]
A users want to add a different tag + value per Retry instance. 

Constraints:
The TaggedRetryMetrics should not have any knowledge which instances are actually registered in the Registry, because instances can be registered dynamically. 

Thus I guess we have to add the custom tags to a specific instance. But we can't add tags to the RetryConfig, because a RetryConfig can be shared by different instances, but still a specific instance should be able to have different tag values. The Retry instance itself is immutable. You can't add tags to an instance after it has been created. I could think of an optional HashMap which can be used to store some metadata to instances.

Like:

```
retryRegistry.of("name", retryConfig, metaDataMap)

```

**Comment by siepkes** [comment]
What your proposing is to allow every individual `Retry` instance to have it's own specific tags? What I was proposing was specific tags for a registry. 

I think modifying `RetryRegistry.of()` to allow passing a map of (un-typed) meta data which will later be used by specific modules like the MicroMeter module does make it a easy to make mistakes which aren't caught by the compiler. I understand the problem; The core module where the `RetryRegistry.of()` API lives knows nothing about MicroMeter so passing a `Tags` object is not an option. 

Personally I feel that API-safety-wise it would be better to allow tags to be specified per repository like this: 

```java
public static TaggedBulkheadMetrics ofBulkheadRegistry(Tags tags, MetricNames names, BulkheadRegistry bulkheadRegistry) { }
```
This allows for the use of MicroMeter specific API in the API. Basically if people need to have specific tags they would then need to create a separate repository. Since they can re-use (I believe?) the scheduled executor I don't think that would be to big of a "cost" with the gain of keeping type info (and a clearer API).

I believe the original poster (@batigoal82 but correct me if I'm wrong) also just wanted per repo specific tags since he talks about the private constructors of classes like `TaggedRetryMetrics`. Even with public constructors you can only add tags to repositories, not to individual `Retry` instances.

Let me know what you think!

**Comment by dlsrb6342** [comment]
What do you think about adding `TagsProvider` interface?
Please check https://github.com/spring-cloud/spring-cloud-gateway/pull/1116 PR. I made the function you want in spring-cloud-gateway.
The advantage of this interface is that it can be added simply when the user wants to add more tags.

**Comment by RobWin** [comment]
I answer you tomorrow in more detail. I only have a smartphone now.
Another constraint is that the solution should be also usable with Spring Boot.
Our Spring Boot AutoConfiguration only creates on registry instance. The user should not be forced to create another instance and should be able to configur the instances via the properties/yaml file.

**Comment by RobWin** [comment]
But does the `TagsProvider` solution also allow the user to add tags via a config file?

**Comment by dlsrb6342** [comment]
@RobWin 
It can support config file. I just made it. 
https://github.com/spring-cloud/spring-cloud-gateway/pull/1311

**Comment by RobWin** [comment]
>I think modifying RetryRegistry.of() to allow passing a map of (un-typed) meta data which will later be used by specific modules like the MicroMeter module does make it a easy to make mistakes which aren't caught by the compiler. 

@siepkes Why do you think the map is `un-typed`? The meta-data map should be Map<String, String>. The Micrometer tags are also `tag(String key, String value)`.

**Comment by RobWin** [comment]
I could imagine to configure tags in Spring Boot as follows:

```
resilience4j.retry:
    configs:
        default:
            maxRetryAttempts: 2
            waitDuration: 100
            tags:
                module: awesomeModule
                tag2: tag2Value
    instances:
        backendA:
            baseConfig: default
            tags:
                module: overwriteModuleName
                tag3: tag3Value
```

It would even allow to overwrite tags/metadata.

**Comment by RobWin** [comment]
I think if we add tags/labels/metadata as an integral part of Resilience4j, we don't need separate solution for every framework like Spring Boot, Micronaut, Ratpack.

A user could retrieve tags with `retry.getTags()` 
and add tags with `retryRegistry.of("name", retryConfig, tags)`

**Comment by siepkes** [comment]
> Why do you think the map is un-typed? The meta-data map should be Map<String, String>. The Micrometer tags are also tag(String key, String value).

@RobWin I assumed the data would be untyped because of the proposed interface. The proposed interface lives in the core module which knows nothing about the MicroMeter module. So that would potentially mean the `metaDataMap` is also allowed to be used by other modules then the MicroMeter module. So therefor I assumed the key would be the name of the module the data belongs to (for example `micrometer`). Otherwise there is no way for modules to tell if the metadata belongs to them. However the MicroMeter tag itself is also a key and a value which would need to be compressed in to a single string then. Either way when using the module name as key using `Map<String, Object>` might work better. Though as I understand it now you were proposing to just put the tags directly in the metadata map?

> I think if we add tags/labels/metadata as an integral part of Resilience4j, we don't need a separate solution for every framework like Spring Boot, Micronaut, Ratpack.

If tags are made a first class citizen in the core API then that would indeed simplify a lot of things. It would alleviate the entire metadata thing. I assume you are proposing creating a simple `Tag` class which lives in resillience4j which can then be converter into for example a MicroMeter `Tag` instance in the micrometer module (or whatever meaning modules give to a tag)?

**Comment by RobWin** [comment]
My vote goes to: Integrate tags as an integral part of Resilience4j.

@storozhukBM @Romeh @dlsrb6342 What is your opinion?

**Comment by RobWin** [comment]
Would you like to implement it @siepkes?

**Comment by dlsrb6342** [comment]
I agree with yo @RobWin. the Resilience4j's Tag means instance metadata right? 
And then, all of metadata will be added micrometer's tag?

**Comment by RobWin** [comment]
Yes

**Comment by siepkes** [comment]
@RobWin I don't want to be a party pooper and if you guys think I need to stop whining then feel free to say so :wink: but I have some doubts if adding tags as an integral part of Resilience4j will fix the issue of MicroMeter specific things per (for example) retry instances "once and for all". 

For example MicroMeter also has the ability to add a description (a String) for each metric. How would such a thing be added? Basically you then get the exact same problem as we are having now. We can't keep adding all those things as integral parts of Resilience4j I assume?

**Comment by RobWin** [comment]
I have two little poopers at home 👶 I can handle that ;)

A Tag im Micrometer cannot have a description.
The Meter itself has a description. The Meters (metrics) are created in the resiliencej4-micrometer module. The module already adds a description to every meter and adds tags like the instance name.

The tagging system which I would like to integrate would allow the user to add his own tags. The tags would be used by Micrometer, Prometheus. But the information could also be used by a Logging System (SLF4j) to add context information to a log statement.

**Comment by siepkes** [comment]
In that context it all sounds good to me! Since I'm the guy asking for it I'm willing willing to implement it. 

You pinged some people to give their blessing so I'll await the all clear before starting on it.

**Comment by RobWin** [comment]
I think you can directly start

**Comment by siepkes** [comment]
@RobWin I've created a PR (#654) in which I have added a tag API to the `Retry` class as discussed. I thought it would be a good idea to check in with you folks to see if I got the API idea right before proceeding.

**Comment by RobWin** [comment]
Could you extend the micrometer, prometheus modules now?

**Comment by siepkes** [comment]
@RobWin Great, thanks! Yeah I'll give MicroMeter and Prometheus the tag treatment. Will probably be somewhere next week before I can get started though.

**Comment by siepkes** [comment]
Just a quick update; I won't be able to start on this this week. With a bit of luck i'll be able to pick this up next week.

**Comment by RobWin** [comment]
When do you think you will have some time?

**Comment by Romeh** [comment]
@RobWin i will add MicroMeter and Prometheus tag treatment.

**Comment by siepkes** [comment]
@Romeh Sorry folks, I got side-tracked by a bunch of issues that....errr....suddenly  became a priority ;-). If you could start on MicroMeter and Prometheus integration that would be great! If you have any questions about how the basic tag support is implemented in resilience4j just let me know!

**Comment by Romeh** [comment]
Ok micrometer tags support is done and merged , spring boot configuration support next to do !

**Comment by springfan** [comment]
Can you guys give an example regarding the usage / has that been documented anywhere?

**Comment by RobWin** [comment]
In Spring Boot:

```
resilience4j.circuitbreaker:
  tags:
    tag1: tag1Value
    tag2: tag2Value
```

**Comment by springfan** [comment]
Thanks. I assume for instance specific tags it works similar?

**Comment by springfan** [comment]
Is there a way to use the Spring boot annotations with a dynamic tag? I would like to add the user as tag to dinstinguish the metric calls by user.

**Comment by dlsrb6342** [comment]
Not supported but it would be nice feature. 
And if you would like more questions or suggestions, please do not write on closed issue and create new issue.

**Comment by RobWin** [comment]
Hi, you should never use a user ID as a tag. If your system must scale to thousands of users, your metric system would be in trouble.
See Prometheus for example: https://prometheus.io/docs/practices/instrumentation/#do-not-overuse-labels

**Comment by springfan** [comment]
Hi. Yes in my case the number of users would be very limited < 30. My initital thought was why should I setup a new metric when resilience4j already measures all the requests. But I think it's better to spin up a seperate metric for API/user calls that does not measure fault tolerance related behavior to limit the cardinality and isolate the use case.
